src/client.ts

```diff
@@ -9297,31 +9297,51 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
 
     /**
      * Find the event_id closest to the given timestamp in the given direction.
-     * @return {Promise} A promise of an object containing the event_id and
-     *    origin_server_ts of the closest event to the timestamp in the given
-     *    direction
+     * @return {Promise} Resolves: A promise of an object containing the event_id and
+     *    origin_server_ts of the closest event to the timestamp in the given direction
+     * @return {module:http-api.MatrixError} Rejects: when the request fails
      */
-    public timestampToEvent(
+    public async timestampToEvent(
         roomId: string,
         timestamp: number,
         dir: Direction,
     ): Promise<ITimestampToEventResponse> {
         const path = utils.encodeUri("/rooms/$roomId/timestamp_to_event", {
             $roomId: roomId,
         });
+        const queryParams = {
+            ts: timestamp.toString(),
+            dir: dir,
+        };
 
-        return this.http.authedRequest(
-            Method.Get,
-            path,
-            {
-                ts: timestamp.toString(),
-                dir: dir,
-            },
-            undefined,
-            {
-                prefix: "/_matrix/client/unstable/org.matrix.msc3030",
-            },
-        );
+        try {
+            return await this.http.authedRequest(
+                Method.Get,
+                path,
+                queryParams,
+                undefined,
+                {
+                    prefix: ClientPrefix.V1,
+                },
+            );
+        } catch (err) {
```

Comment:
Need exception for SonarCloud, https://sonarcloud.io/dashboard?id=matrix-js-sdk&pullRequest=2915

No interest in making a test for testing the prefixed fallback. It's all mocked garbage anyway.

Reply:
Why not instead check the versions endpoint to see which spec releases the server supports and deduce API support level that way? One less round-trip per request and more testable.

Reply:
@t3chguy `/timestamp_to_event` isn't in a Matrix spec version yet to deduce from and the stable version is not in `unstable_features`.

Reply:
So how is the client meant to know whether to show the feature to the user or not? Normally unstable_features includes both the unstable and stable-but-not-in-a-spec-release flags.

https://github.com/matrix-org/synapse/blob/13aa29db1ddc925beb35f5f1da8fd1a1bcc91373/synapse/rest/client/versions.py#L82-L123

Reply:
fwiw, `stable-but-not-in-spec-release` flags are only meant to be created in two scenarios:
1. The stable release of the feature is expected to span multiple spec releases
2. The feature cannot be detected any other way (4xx error, etc)

We (the SCT) generally don't like to assign such flags because they're confusing and complicated to manage at the implementation level.

Instead, the recommendation is to convert the now-stable MSC to actual spec *before* updating the implementations to help build confidence in the spec version the changes will land in, allowing clients to do a simpler check for stability. In this case however, it appears the implementations are being updated first, which can cause process problems.

With my SCT hat on, I'd recommend pausing work on this PR and instead refocusing on writing up the MSC as actual spec. Currently the SCT is scheduled to look at it, so please drop by [#sct-office:matrix.org](https://matrix.to/#/#sct-office:matrix.org) to confirm if we're too far along the spec writing process for it to reasonably be taken over (the SCT's latest shift in priorities does not include writing up spec PRs for MSCs, so if the author or another person is more interested, we'll generally stop working on it in favour of someone else picking it up).

Reply:
> Normally unstable_features includes both the unstable and stable-but-not-in-a-spec-release flags.

I asked about "advertising support" in https://github.com/matrix-org/synapse/issues/14390. Using `org.matrix.msc3030.stable` would work well until we get a spec release 👍. Thanks for the example! I'll create a Synapse PR for that.

The alternative to determine support is to run `timestampToEvent(...)` against one of the users rooms and see if it succeeds or gives us a `M_UNRECOGNIZED`. If they haven't joined a room, then we just assume no support (small edge case until we get a spec release).


For reference, [`getRoomHierarchy`](https://github.com/matrix-org/matrix-js-sdk/blob/6611cfa2537ef14ce8c8f7d76a45dad4a4e26167/src/client.ts#L9116-L9146) uses this fallback pattern.

---

@turt2live We're supposed to be able to move fast with all of these things. Even if the spec PR was written, it seems like your suggestion is to wait for a Matrix spec release which isn't the fast thing.

It's unclear to me what your suggestion for moving fast is at least.

Reply:
> The feature cannot be detected any other way (4xx error, etc)

This is fine for deciding which API to use, but not for whether or not to show the feature at all, having to call it with dummy params may fail due to stringent input checking on the server

Reply:
> @turt2live We're supposed to be able to move fast with all of these things. Even if the spec PR was written, it seems like your suggestion is to wait for a Matrix spec release which isn't the fast thing.
> 
> It's unclear to me what your suggestion for moving fast is at least.

By opening the spec PR, we'd move it from "will probably get written up eventually" to "next release material", allowing you to consider what that version number would be. The spec PR would also have to understand which spec release it's going into for the `added-in` flags sprinkled throughout the spec. 

Writing the spec PR first is the faster approach compared to managing 2 unstable flags and a spec release.

Reply:
Quarterly spec releases is not fast for progressively using features.

I could have the stable endpoint used tomorrow using `org.matrix.msc3030.stable`.

Reply:
In case it's not clear, I'm strongly advocating *against* adding a `stable-but-unreleased` flag as it ultimately causes more problems than it solves for a feature of this size.

Reply:
> Quarterly spec releases is not fast for progressively using features.
> 
> I could have the stable endpoint used tomorrow using `org.matrix.msc3030.stable`.

is there a reason why you need the feature before a spec release? The unstable endpoint should be just as sufficient for this feature?

fwiw, the SCT is completely unaware of your timelines or priorities on this. If there's something driving this work which demands a faster timeline than the one available, we need to know by having it raised in the SCT office linked above.

Reply:
> is there a reason why you need the feature before a spec release? The unstable endpoint should be just as sufficient for this feature?

The unstable endpoint is gone in the next Synapse release, https://github.com/matrix-org/synapse/pull/14471 (only stable endpoint)

> fwiw, the SCT is completely unaware of your timelines or priorities on this. If there's something driving this work which demands a faster timeline than the one available, we need to know by having it raised in the SCT office linked above.

We can move fast on features outside of the SCT. Nothing is needed from the SCT end for this to work.

The drive to do this work is to get it done so people can enjoy the feature when `matrix.org` gets the stable endpoint in the next Synapse release. Get a feature out the door.

I'm aware you want the spec PR written from [MSC3030](https://github.com/matrix-org/matrix-spec-proposals/pull/3030) but that's independent of us being able to accomplish things now.

Reply:
> > is there a reason why you need the feature before a spec release? The unstable endpoint should be just as sufficient for this feature?
> 
> The unstable endpoint is gone in the next Synapse release, [matrix-org/synapse#14471](https://github.com/matrix-org/synapse/pull/14471) (only stable endpoint)
> 

This feels premature and like a mistake, considering the impact here :(

The unstable and stable endpoints provide the exact same functionality (they did not change between implementation and acceptance of the MSC), and it would be more valuable to the end user to continue using the feature while the spec process does its thing: the unstable endpoint is more than able to be used for this.

Releasing the feature does *not* have to mean using stable endpoints.

> > fwiw, the SCT is completely unaware of your timelines or priorities on this. If there's something driving this work which demands a faster timeline than the one available, we need to know by having it raised in the SCT office linked above.
> 
> We can move fast on features outside of the SCT. Nothing is needed from the SCT end for this to work.
> 
> The drive to do this work is to get it done so people can enjoy the feature when `matrix.org` gets the stable endpoint in the next Synapse release. Get a feature out the door.
> 
> I'm aware you want the spec PR written from [MSC3030](https://github.com/matrix-org/matrix-spec-proposals/pull/3030) but that's independent of us being able to accomplish things now.

The SCT does need to be involved in consideration for unstable feature flags like this, as it affects the MSC's overall timeline and text. Adding a flag after acceptance is a process burden, as it needs to be formally documented and could even be rejected.

Aside from the SCT's involvement, the product and design teams also need to be consulted on this if you're aiming to release the feature as it's currently behind labs. Labs features are intended to break from time to time (so it's feasible to remove the unstable endpoint if you like, though expectations of having a stable endpoint can cause users to lose access to the feature), and removing those features from labs requires explicit sign off: the feature wouldn't be "shipped" just because a stable endpoint is being used.

Reply:
> This feels premature and like a mistake, considering the impact here :(
> 
> The unstable and stable endpoints provide the exact same functionality (they did not change between implementation and acceptance of the MSC), and it would be more valuable to the end user to continue using the feature while the spec process does its thing: the unstable endpoint is more than able to be used for this.
> 
> Releasing the feature does not have to mean using stable endpoints.

It won't be enabled on `matrix.org` unless it's stable. With the next Synapse release, it will be stable.

This means a lot of people can actually use the jump to date lab on Element if they want.

We also need `/timestamp_to_event` enabled for the [canonical instance of the Matrix Public Archive](https://github.com/matrix-org/matrix-ansible-private/issues/5325) which means having it enabled on `matrix.org`.

As I mentioned [here](https://github.com/matrix-org/synapse/pull/14471#discussion_r1024523025), I'm not worried if there is a period where that functionality disappears. And there doesn't even have to be a period where it disappears if we merge something like this PR.


> Aside from the SCT's involvement, the product and design teams also need to be consulted on this if you're aiming to release the feature as it's currently behind labs. Labs features are intended to break from time to time (so it's feasible to remove the unstable endpoint if you like, though expectations of having a stable endpoint can cause users to lose access to the feature), and removing those features from labs requires explicit sign off: the feature wouldn't be "shipped" just because a stable endpoint is being used.

Whatever product/design input that needs to happen can happen on the de-labs PR for the jump to date feature.


---

I still see no alternative suggested that provides an equivalent speed to iteration for shipping a feature.

Reply:
To bring the quote wall back into a single thread: I'm trying to point out that there's nothing being shipped here. It's completely valid to have *both* unstable and stable endpoints supported by Synapse *without a config flag to disable the unstable feature* - if the stable endpoint is exposed, that should be more than enough reason to expose the unstable one too. 

The Synapse team's restrictions are also managed more easily when going through the spec writing path rather than the feature flag path. It's considerably less effort to write the spec up, get a known spec release assigned to the feature, and check for that spec release in client code (while still using the unstable endpoint until spec release) than it is to try and manage the heavyweight process of a stable feature flag.

This also applies to the Matrix Public Archive project and any other implementation which we might not be aware of. 

I'm not sure what planning has been done so far, but it really feels like this discussion is rooted in misinterpretation of how all the different processes interact and their purposes. Please also understand that I'm not making these suggestions to send you down the slowest or most inconvenient possible process: I genuinely believe that there is a better (and faster) way to get the feature *fully released* than the plan suggested by this PR.

Reply:
on this specific point though:

> And there doesn't even have to be a period where it disappears if we merge something like this PR.

Not everyone uses matrix.org, the latest Synapse release, or the latest Element release. Users are ultimately affected more than usual by the path this PR (and the associated Synapse PR) take.

Reply:
Further, while we've been having this conversation, Hubert has been working away at writing the spec PR - much of the discussion is effectively moot now that it's formally on the table for the next spec release.

Reply:
> It's completely valid to have *both* unstable and stable endpoints supported by Synapse *without a config flag to disable the unstable feature* - if the stable endpoint is exposed, that should be more than enough reason to expose the unstable one too. 

Leaving the unstable endpoint in place until a spec release happens is a reasonable option 👍 - I can create a PR to add it back.

But I'd still like to move forward with a change like this PR so we can also use `/timestamp_to_event` on `matrix.org` ahead of the spec release (whether that includes `org.matrix.msc3030.stable` or however).



> Further, while we've been having this conversation, Hubert has been working away at writing the spec PR - much of the discussion is effectively moot now that it's formally on the table for the next spec release.

Having a spec PR is fantastic!

But this involves a lot more than a spec PR. We have to wait for the next Matrix spec release to be cut, we have to create an issue in Synapse to track changes for that new spec version, we have to implement everything, then advertise support for that spec version.

(these things aren't fast)

There is no question that this is the completely stable future but there is space to move without it.

Reply:
Reinstating the unstable endpoint means there's no need for an unstable-stable flag, as implementations can simply use the unstable endpoint. We know from the spec PR that it'll be in v1.6 of Matrix, so we can write client code to detect that specific version and use the stable endpoint in that case, which will automatically take effect when the server supports the spec version.

When the server adds support for advertising the spec version, it can also remove the unstable endpoint, leaving the feature fully functional for everyone. On matrix.org, the unstable endpoint would also be available with this setup, meaning a larger testing group for the Product team's consideration.

The feature can be de-labsed (released) at any point from here on out - with or without a spec release happening, and with or without a stable endpoint to call.

While these steps individually are not fast, when considered as a whole they are extremely efficient as they are designed for this exact scenario: wanting to get a feature live *before* the spec release even happens. An unstable-stable feature flag ends up *adding* steps, which means additional workload for all the teams involved and therefore more time.

Regardless, if you'd like to still continue with this PR, tests will be required per my other thread/review. I'd suggest adapting the PR to meet the plan I've outlined above, but ultimately the concern is lack of tests rather than the approach (as hopefully highlighted by the formal review left with my web team hat on).

Reply:
> Reinstating the unstable endpoint means there's no need for an unstable-stable flag, as implementations can simply use the unstable endpoint. We know from the spec PR that it'll be in v1.6 of Matrix, so we can write client code to detect that specific version and use the stable endpoint in that case, which will automatically take effect when the server supports the spec version.

Thinking about this more, while it's still reasonable to leave the unstable endpoint alongside the stable one, I don't want to maintain both. This involves complexity on the server to choose the right federation endpoint depending on the entry and if we want a seamless transition while we support both, we have to do some fallbacks on the server as well. And underlying everything is that we don't have to let breaking unstable and labs features slow us down (they have no guarantees).

Things flow just fine on the client if we use the stable endpoint and fallback to unstable one.

For deciding whether to show a feature or not, I'd even lean towards just calling `timestampToEvent(firstRoom, 0, 'f')` and seeing if we get a successful response or an `M_UNRECOGNIZED` error. This avoids the unstable-stable flag as well.

Then when Matrix v1.6 is out, we can remove the quick and dirty support check in favor of a `v1.6` check. We could also do this first and fallback to the quick and dirty check.


> While these steps individually are not fast, when considered as a whole they are extremely efficient as they are designed for this exact scenario: wanting to get a feature live before the spec release even happens. An unstable-stable feature flag ends up adding steps, which means additional workload for all the teams involved and therefore more time.

No denying that waiting for every piece to be in place is the most efficient and the perfect thing to do if you have time to wait around 👍

Waiting for Matrix spec release:

 1. PR for unstable implementation on server
 1. PR for unstable implementation on client
 1. PR for stable implementation on server (dropping unstable implementation)
 1. PR for stable implementation on client with fallback to unstable (because easy)
 1. PR to remove unstable fallback on client

But that actually turns out to be almost the same amount of work as I'm trying to approach things. We just lose a little bit of compatibility in the transition:

 1. PR for unstable implementation on server
 1. PR for unstable implementation on client
 1. PR for stable implementation on server (dropping unstable implementation)
 1. PR for stable implementation on client with fallback to unstable (because easy)
 1. PR to remove unstable fallback and any updates for final stable checks on client

And if we try to be the most compatible by supporting both unstable and stable at the same time, it's actually more work:

 1. PR for unstable implementation on server
 1. PR for unstable implementation on client
 1. PR for stable implementation on server (with complexity of supporting both)
 1. PR for stable implementation on client with fallback to unstable
 1. PR to remove unstable implementation from server
 1. PR to remove unstable fallback on client

(hopefully these are right but it's kinda a whatever point. I don't disagree basically)


> Regardless, if you'd like to still continue with this PR, tests will be required per my other thread/review. I'd suggest adapting the PR to meet the plan I've outlined above, but ultimately the concern is lack of tests rather than the approach (as hopefully highlighted by the formal review left with my web team hat on).

👍 Sounds good, I'll add something here ⏩

Reply:
> I'd also recommend just trying the real request instead of a limit of `0` - worst case is you look for [MSC3743](https://github.com/matrix-org/matrix-spec-proposals/pull/3743)'s error codes and repeat the same exact request.

For reference, `timestampToEvent(firstRoom, 0, 'f')` is a real request and `0` is the timestamp in this example. It will find the first event in the room.

Looks like we're thinking the same on all things around the check though 👍


> fwiw, in an unexpected discussion with the Synapse team around reinstating the unstable endpoint for this MSC in particular, the team is highly resistant to enable it by default due to factors I wasn't aware of. Those factors are being considered externally, but not in time for this PR's future.

Where did this discussion happen (link please, even if internal)? I'm looking for more context and what the factors are.

Reply:
> > fwiw, in an unexpected discussion with the Synapse team around reinstating the unstable endpoint for this MSC in particular, the team is highly resistant to enable it by default due to factors I wasn't aware of. Those factors are being considered externally, but not in time for this PR's future.
> 
> Where did this discussion happen (link please, even if internal)? I'm looking for more context and what the factors are.

It's specifically internal to the Spec Core Team at the moment. The results of the conversation will be made public as needed.